Sources/SwiftLayout/Refactoring/Protocol/LayoutTraversal.swift

```diff
@@ -0,0 +1,201 @@
+import Foundation
+import UIKit
+
+typealias TraverseHandler = (_ superview: UIView?, _ subview: UIView, _ identifier: String?, _ animationDisabled: Bool) -> Void
+typealias ConstraintHandler = (_ superview: UIView?, _ subview: UIView, _ constraints: [Constraint], _ viewInfoSet: ViewInformationSet) -> Void
+
+protocol LayoutTraversal {
+    func traverse(_ superview: UIView?, continueAfterViewLayout: Bool, traverseHandler handler: TraverseHandler)
+    func traverse(_ superview: UIView?, viewInfoSet: ViewInformationSet, constraintHndler handler: ConstraintHandler)
+}
+
+extension Array where Element == LayoutTraversal {
+    mutating func appendLayout<L: Layout>(_ l: L) {
+        guard let traversal = l as? LayoutTraversal else { return }
+        append(traversal)
+    }
+}
+
+extension LayoutTraversal {
+    var viewInformations: [ViewInformation] {
+        var informations: [ViewInformation] = []
+        traverse(nil, continueAfterViewLayout: true) { superview, subview, identifier, animationDisabled in
+            informations.append(.init(superview: superview, view: subview, identifier: identifier, animationDisabled: animationDisabled))
+        }
+        return informations
+    }
+    func viewConstraints(_ viewInfoSet: ViewInformationSet) -> [NSLayoutConstraint] {
+        var layoutConstraints: [NSLayoutConstraint] = []
+        traverse(nil, viewInfoSet: viewInfoSet) { superview, subview, constraints, viewInfoSet in
+            layoutConstraints.append(contentsOf: constraints.constraints(item: subview, toItem: superview, viewInfoSet: viewInfoSet))
+        }
+        return layoutConstraints
+    }
+    
+    func cast<L: Layout>(_ layout: L) -> LayoutTraversal? {
+        layout as? LayoutTraversal
+    }
+}
+
+extension AnchorsLayout: LayoutTraversal {
+    func traverse(_ superview: UIView?, continueAfterViewLayout: Bool, traverseHandler handler: TraverseHandler) {
```

Comment:
What do you think about putting extensions on each concrete layouts?

Reply:
agreed